[PGO] Sampled instrumentation in PGO to speed up instrumentation binary

xur-llvm opened the pull request:
In comparison to non-instrumented binaries, PGO instrumentation binaries
can be significantly slower. For highly threaded programs, this slowdown can
reach 10x due to data races or false sharing within counters.

This patch incorporates sampling into the PGO instrumentation process to
enhance the speed of instrumentation binaries. The fundamental concept
is similar to the one proposed in https://reviews.llvm.org/D63949.

Three sampling modes are introduced:
1. Simple Sampling: When '-sampled-instr-bust-duration' is set to 1.
2. Fast Burst Sampling: When not using simple sampling, and 
    '-sampled-instr-period' is set to 65535. This is the default mode of sampling.
3. Full Burst Sampling: When neither simple nor fast burst sampling is used.

Utilizing this sampled instrumentation significantly improves the binary's
execution speed. Measurements show up to 5x speedup with default
settings. Fast burst sampling now results in only around 20% to 30%
slowdown (compared to 8 to 10x slowdown without sampling).

Out tests show that profile quality remains good with sampling,
with edge counts typically showing more than 90% overlap.
For applications whose behavior changes due to binary speed,
sampling instrumentation can enhance performance.
Observations have shown some apps experiencing up to
a ~2% improvement in PGO.

A potential drawback of this patch is the increased binary size
and compilation time. The Sampling method in this patch does
not improve single threaded program instrumentation binary
speed.

xur-llvm committed:
Sampled instrumentation in PGO to speed up instrumentation binary

PGO instrumentation binary can be very slow comparing to the
non-instrumented binary. It's not uncommon to see 10x slowdown
for highly threaded programs, due to data race of false sharing
to the counters.

This patch uses sampling in PGO instrumentation to speed up
instrumentation binary. The basic idea is the same as one:
here: https://reviews.llvm.org/D63949

This patch makes some improvements so that we only use one
condition. We now fix the WholeDuring at 65536 and use the
wraparound of unsigned short.

With this sampled instrumentation, the binary runs much
faster. We measure 5x speedup using the default duration.
We now only see about 20% to 30% slow down (comparing to
8 to 10x slowdown without sampling).

The profile quality is pretty good with sampling: the edge
counts usually report >90% overlap.

For the apps that program behaviors change due to binary
speed, sampling instrumentation can improve the performance.
We have observed some apps getting up ~2% improvement in PGO.

One potential issue of this patch is the increased binary
size and compilation time.

snehasish commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -805,6 +808,10 @@ void PassBuilder::addPGOInstrPasses(ModulePassManager &MPM,
   Options.UseBFIInPromotion = IsCS;
+  if (EnableSampledInstr) {
+    Options.Sampling = true;
+    Options.DoCounterPromotion = false;
```
Can you add a comment on why counter promotion is turned off?

xur-llvm replied:
The reason is mentioned in InstrProfling.C:400.

snehasish commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+// We transform:
+//   Increment_Instruction;
+// to:
+//   if (__llvm_profile_sampling__ <= SampleDuration) {
+//     Increment_Instruction;
+//   }
+//   __llvm_profile_sampling__ += 1;
```
This seems more like `firstN of 64K` rather than random sampling. I have a couple of concerns.
1. This seems like it could easily introduce bias.
2. We introduce yet another knob for tuning `sampled-instr-duration`. If it was randomized then we won't have to worry about this knob. A fast random implementation may not be too much increase in overhead compared to this approach.

WenleiHe replied:
Agree with what @snehasish said. Why not record once for every N hits? firstN of 64K is more prone to synchronization issues (with the program).

david-xl replied:
xur@'s sampling scheme is bursty style which seems less likely to introduce bias compared with every N hits scheme (which can lead to shadow. 

For this style of sampling, there might be some optimization that can be done to coalesce sample count update and check.

htyu replied:
Since `__llvm_profile_sampling__` is shared by all PGO counters, I'm wondering if biasing could happen to code always shadowed in the range `[__llvm_profile_sampling__, 65536]`. E.g, assume two loops in the program, they count up to 65536 iterations in all, and the second loop will unlikely be counted?

david-xl replied:
It can be problem for static count, but might be ok with runtime count. Data changes at runtime introduce some randomness -- for instance loop trip count. It is unlikely all loops in a program have fixed trip count through out the training run.

xur-llvm replied:
__llvm_profile_sampling__ is not shared by all PGO counters. It's s thread-local variable. All the counters in one thread shared the value.

This is burst sampling. We used to have to parameters (i.e. changing 64K to another used specified value).  Using 64K value does increase the chances of bias. But I do think the chances are low in real programs.

Of course, I can write a test case to show the method will result in a bias profile. This is sampling after all. There will be always corner cases that gets biased result.

WenleiHe replied:
> xur@'s sampling scheme is bursty style which seems less likely to introduce bias compared with every N hits scheme

Why? In general, sampling needs to be evenly distributed to avoid bias. HW sampling doesn't do burst either.  Do you have data and example to illustrate the benefit of bursty sampling? 

The only reason I can see to choose bursty sample is to open up the opportunity for coalescing increments, but that is not done here.

snehasish commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+      MadeChange = true;
+    } else {
+      LLVM_DEBUG(dbgs() << "Invalid InstroProf intrinsic: " << *IP << "\n");
+      // ?? Seeing "call void @llvm.memcpy.p0.p0.i64..." here ??
+      // llvm_unreachable("Invalid InstroProf intrinsic");
```
I think the llvm_unreachable should be uncommented?

xur-llvm replied:
I think there is bug: it seems I get llvm.memcpy instinsic here. That's the reason I comment out and sink the doSampling(IP) to the loop. I had this patch for quite a while. I'll try to see if the bug still there.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -148,6 +150,16 @@ cl::opt<bool> SkipRetExitBlock(
+                      cl::desc("Do PGO instrumentation sampling"));
+
+static cl::opt<unsigned> SampledInstrumentDuration(
+    "sampled-instr-duration",
```
nit: this is more like a sample rate rather than duration?

david-xl replied:
or sampling period.

xur-llvm replied:
I don't this is the sampling rate. This is burst sampling. In this burst period, the sample rate is 1. this parameter specifies the during for the burst period.

WenleiHe replied:
If you end up with bursty sampling, maybe `sampled-instr-burst-duration` to be clear. There also needs to be a explanation in comment somewhere to call out burst mode is chosen and explain why.

It can be useful to have a separate `sampled-instr-period`, which defaults to 64K, so that part is also tunable when needed. With that flag, people can set `sampled-instr-burst-duration` to 1, and tune `sampled-instr-period` instead to achieve non-burst sampling. We may give it a try.

snehasish replied:
I like the idea of two flags and allowing users to choose the setting enabling to both bursty or conventional sampling. One suggestion would be to make `sampled-instr-period` a prime number (standard practice in hardware sampling).

david-xl replied:
The current sampling period is implicit to take advantage of  the wrapping behavior to reduce runtime overhead (and code size) -- we don't want to lose that. However I think the implementation can check if the bursty duration is not one, assert that sampling period is not set. Otherwise, generate the non bursty style sampling.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+  Instruction *ThenTerm = SplitBlockAndInsertIfThen(
+      DurationCond, I, /* Unreacheable */ false, BranchWeight);
+  IRBuilder<> IncBuilder(I);
+  auto *NewVal = IncBuilder.CreateAdd(LoadCountVar, IncBuilder.getInt16(1));
```
So the wrap around is implicitly dependent on HW implementation of integer add?

xur-llvm replied:
Not sure I get the question: this depends on the wrap-around of unsigned add. I thought this is a c/c++ standard.

WenleiHe replied:
You're right, signed add overflow is UB, but overflow of unsigned add should always wrap around per standard.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
 
+  for (auto *IP : InstrProfInsts) {
+    if (auto *IPIS = dyn_cast<InstrProfIncrementInstStep>(IP)) {
+      doSampling(IP);
```
nit: hoist `doSampling(IP)` out of the if-else construct?

xur-llvm replied:
This is to work-around the bug in line 507.

WenleiHe reviewed (commented):
> One potential issue of this patch is the increased binary size and compilation time.

Typically how much bigger is the .text size?

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+//   __llvm_profile_sampling__ += 1;
+//
+// "__llvm_profile_sampling__" is a thread-local global shared by all PGO
+// instrumentation variables (value-instrumentation and edge instrumentation).
```
instrumentation variables --> counters

xur-llvm replied:
Name is shared by all counters. I will change the comments.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+// It has a unsigned short type and will wrapper around when overflow.
+//
+// Note that, the code snippet after the transformation can still be
+// counter promoted. But I don't see a reason for that because the
```
I don't see a reason --> there is no reason ..

xur-llvm replied:
ack.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+// Note that, the code snippet after the transformation can still be
+// counter promoted. But I don't see a reason for that because the
+// counter updated should be sparse. That's the reason we disable
+// counter promotion by default when sampling is enabled.
```
It is still unclear why counter promotion should be disabled with this comment.

xur-llvm replied:
OK. I can expand a bit more there: The downside of counter promotion is that we can get incomplete profile if we dump the counter in the middle of the loop. The benefit is improve the instrumentation speed. With this patch, the benefit is very small and won't overweight the potential downside.

As I said in the comment, these two techniques can work together without any issue. I actually added a test case for them to work together.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -148,6 +150,16 @@ cl::opt<bool> SkipRetExitBlock(
     cl::desc("Suppress counter promotion if exit blocks contain ret."));
 
+static cl::opt<bool>
+    SampledInstrument("sampled-instr", cl::ZeroOrMore, cl::init(false),
```
naming nit: SampledInstr to be consistent with command name. Similar change in other places.

xur-llvm replied:
OK. Will change.

snehasish commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -412,30 +424,91 @@ PreservedAnalyses InstrProfiling::run(Module &M, ModuleAnalysisManager &AM) {
+  MDNode *BranchWeight =
+      MDB.createBranchWeights(SampleDuration, WrapToZeroValue - SampleDuration);
+  Instruction *ThenTerm = SplitBlockAndInsertIfThen(
+      DurationCond, I, /* Unreacheable */ false, BranchWeight);
```
nit: typo Unreachable

WenleiHe commented:
Any follow ups on this patch? Once it's in we'd like to give it a try as well.

xur-llvm commented:
Hi, Wenlei,

I just updated the patch to sync with LLVM head. You can try with option
"--mllvm -enable-sampled-instr=true".
I resumed work on this from the last few weeks and I'm testing it with some
internal benchmarks.

One thing that we noticed is that this patch can increase the text size
quite a bit and sometimes hits the 2GB relocation limit for large programs.
I am thinking of doing a function level sampling so that the text size
increase would be minimal.

Please let me know if you have any problems with the patch and
keep us updated about the performance.

-Rong


On Fri, Jan 19, 2024 at 6:51 PM wenlei ***@***.***> wrote:

> Any follow ups on this patch? Once it's in we'd like to give it a try as
> well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/69535#issuecomment-1901669781>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XXW4DZVXN42HYPHGUTYPMWKZAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBRGY3DSNZYGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

WenleiHe commented:
> Hi, Wenlei, I just updated the patch to sync with LLVM head. You can try with option "--mllvm -enable-sampled-instr=true". I resumed work on this from the last few weeks and I'm testing it with some internal benchmarks. One thing that we noticed is that this patch can increase the text size quite a bit and sometimes hits the 2GB relocation limit for large programs. I am thinking of doing a function level sampling so that the text size increase would be minimal. Please let me know if you have any problems with the patch and keep us updated about the performance.
> […](#)
> -Rong
> On Fri, Jan 19, 2024 at 6:51 PM wenlei ***@***.***> wrote: Any follow ups on this patch? Once it's in we'd like to give it a try as well. — Reply to this email directly, view it on GitHub <[#69535 (comment)](https://github.com/llvm/llvm-project/pull/69535#issuecomment-1901669781)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOI42XXW4DZVXN42HYPHGUTYPMWKZAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBRGY3DSNZYGE> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Thanks. Will do!

WenleiHe commented on llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp:
```diff
@@ -770,7 +770,7 @@ BasicBlock *FuncPGOInstrumentation<Edge, BBInfo>::getInstrBB(Edge *E) {
     // There are basic blocks (such as catchswitch) cannot be instrumented.
     // If the returned first insertion point is the end of BB, skip this BB.
-    if (BB->getFirstInsertionPt() == BB->end())
+    if (BB->getFirstNonPHIOrDbgOrAlloca() == BB->end())
```
what is the reason to instrument after alloca?

WenleiHe commented:
`split-indirectbr-critical-edges.ll` need to be updated

WenleiHe commented:
> Hi, Wenlei, I just updated the patch to sync with LLVM head. You can try with option "--mllvm -enable-sampled-instr=true". I resumed work on this from the last few weeks and I'm testing it with some internal benchmarks. One thing that we noticed is that this patch can increase the text size quite a bit and sometimes hits the 2GB relocation limit for large programs. I am thinking of doing a function level sampling so that the text size increase would be minimal. Please let me know if you have any problems with the patch and keep us updated about the performance.
> […](#)
> -Rong
> On Fri, Jan 19, 2024 at 6:51 PM wenlei ***@***.***> wrote: Any follow ups on this patch? Once it's in we'd like to give it a try as well. — Reply to this email directly, view it on GitHub <[#69535 (comment)](https://github.com/llvm/llvm-project/pull/69535#issuecomment-1901669781)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOI42XXW4DZVXN42HYPHGUTYPMWKZAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBRGY3DSNZYGE> . You are receiving this because you authored the thread.Message ID: ***@***.***>

We're seeing ~0.4% rps improvements on HHVM (PHP JIT) with this change on top of regular IRPGO. The smaller improvement is perhaps expected as JIT doesn't have a lot of parallelism/contention.

Any plan for landing this after addressing remaining code review comments? We'd happily take it down after it's landed here -- thanks for the work.

WenleiHe commented:
@xur-llvm is this abandoned?

xur-llvm commented:
On Tue, May 21, 2024 at 7:59 PM wenlei ***@***.***> wrote:

> @xur-llvm <https://github.com/xur-llvm> is this abandoned?
>
No. This is still active. I was busy with other projects in the past few
months. But now I think I have time for this.
I will address the review comments  and land the change.

One issue I found for this patch is that it increases the code size
non-trivially, and it is probamaticaly for certain applications.
We have some ideas to reduce the size. This would be in a follow-up patch.




> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/69535#issuecomment-2123779996>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XRP2IA7AI6N4KEV4N3ZDQC2BAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRTG43TSOJZGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WenleiHe commented:
> On Tue, May 21, 2024 at 7:59 PM wenlei ***@***.***> wrote: @xur-llvm <https://github.com/xur-llvm> is this abandoned?
> No. This is still active. I was busy with other projects in the past few months. But now I think I have time for this. I will address the review comments and land the change. One issue I found for this patch is that it increases the code size non-trivially, and it is probamaticaly for certain applications. We have some ideas to reduce the size. This would be in a follow-up patch.
> […](#)
> — Reply to this email directly, view it on GitHub <[#69535 (comment)](https://github.com/llvm/llvm-project/pull/69535#issuecomment-2123779996)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AOI42XRP2IA7AI6N4KEV4N3ZDQC2BAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRTG43TSOJZGY> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Any update on this @xur-llvm ?

xur-llvm commented:
reopen this pull-request.

xur-llvm commented:
Reopen.

xur-llvm commented:
The last commit was to rebase to the more recent sources.

xur-llvm committed:
[PGO] Sampled instrumentation in PGO to speed up instrumentation binary

Integrated the reviews from Wenlei, David, Snehasish and Hongtao.
The patch now has 3 modes for sampling:
(1) full burst sampling
(2) fast burst sampling
(3) simple sampling
Also update the tests.

xur-llvm commented:
Commit 2 integrated the suggestions from reviewers.
It has full-burst, fast-burst, and simple sampling now.

xur-llvm commented:
I updated the patch yesterday. The new version addressed the review
comments. Could you take a look?

On Thu, Jul 11, 2024 at 10:55 AM wenlei ***@***.***> wrote:

> On Tue, May 21, 2024 at 7:59 PM wenlei *@*.
>
>
> *> wrote: @xur-llvm <https://github.com/xur-llvm>
> https://github.com/xur-llvm <https://github.com/xur-llvm> is this
> abandoned? No. This is still active. I was busy with other projects in the
> past few months. But now I think I have time for this. I will address the
> review comments and land the change. One issue I found for this patch is
> that it increases the code size non-trivially, and it is probamaticaly for
> certain applications. We have some ideas to reduce the size. This would be
> in a follow-up patch. … <#m_3115587418209969750_> — Reply to this email
> directly, view it on GitHub <#69535 (comment)
> <https://github.com/llvm/llvm-project/pull/69535#issuecomment-2123779996>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AOI42XRP2IA7AI6N4KEV4N3ZDQC2BAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRTG43TSOJZGY
> <https://github.com/notifications/unsubscribe-auth/AOI42XRP2IA7AI6N4KEV4N3ZDQC2BAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRTG43TSOJZGY>
> . You are receiving this because you were mentioned.Message ID: @.*>
>
> Any update on this @xur-llvm <https://github.com/xur-llvm> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/69535#issuecomment-2223542537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XR7STCDHS37KOFXG3DZL3BINAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMRTGU2DENJTG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

xur-llvm committed:
[PGO] Fix the test for windows build

Fix the test for windows build as the llvm_nm has different size.
Also rename the tests to correct a typo.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -170,6 +170,29 @@ cl::opt<bool> SkipRetExitBlock(
     "skip-ret-exit-block", cl::init(true),
     cl::desc("Suppress counter promotion if exit blocks contain ret."));
 
+static cl::opt<bool> SampledInstr("sampled-instr", cl::ZeroOrMore,
```
'instr' can be confused with 'instruction'. We should just spell it out as 'sampled-instrumentation'.

xur-llvm replied:
ack.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -170,6 +170,29 @@ cl::opt<bool> SkipRetExitBlock(
+static cl::opt<unsigned> SampledInstrPeriod(
+    "sampled-instr-period",
+    cl::desc("Set the profile instrumentation sample period. For each sample "
+             "period, the 'sampled-instr-burst-duration' number of consecutive "
```
A fixed number of consecutive samples will be record. The number is controlled by 'sampled-instr-burst-duration' flag.

xur-llvm replied:
ack.

david-xl commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -635,33 +664,161 @@ PreservedAnalyses InstrProfilingLoweringPass::run(Module &M,
+// counters (value-instrumentation and edge instrumentation).
+//
+// (2) Fast burst sampling:
+// The value is an unsigned type, meaning it will wrap around to zero when
```
__llvm_profile_sampling__ variable is ..

xur-llvm replied:
ack.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -635,33 +664,161 @@ PreservedAnalyses InstrProfilingLoweringPass::run(Module &M,
+
+  unsigned SampledBurstDuration = SampledInstrBurstDuration.getValue();
+  unsigned SampledPeriod = SampledInstrPeriod.getValue();
+  assert(SampledBurstDuration < SampledPeriod);
```
Should we emit an error? assertion is probably not enough for validating user input.

xur-llvm replied:
Ack. Will change to an error.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -635,33 +664,161 @@ PreservedAnalyses InstrProfilingLoweringPass::run(Module &M,
+  assert(SampledBurstDuration < SampledPeriod);
+  bool UseShort = (SampledPeriod <= USHRT_MAX);
+  bool IsSimpleSampling = (SampledBurstDuration == 1);
+  bool IsFastSampling = (!IsSimpleSampling && SampledPeriod == 65535);
```
My understanding is fast means we don't need a check for period, and instead rely on overflow. In that case, value of SampledBurstDuration is unrelated?

xur-llvm replied:
The value of SampleBurstDuration is number of samples being recored for each duration.  The value will be used in the condition generated.

xur-llvm replied:
The condition is to handle the case where both SampledBurstDuration==1 and SampledPeriod=65535.
I chose to have SimpleSamping over FastSampling. This is just a implementation choice.
Note that we need one condition for the increment with either choices. The difference is either using
the first condition (FastBurst style), or the second condition (simple sampling style).  I don't think that
matters.

WenleiHe commented on llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp:
```diff
@@ -635,33 +664,161 @@ PreservedAnalyses InstrProfilingLoweringPass::run(Module &M,
-        MadeChange = true;
-      }
+      if (auto *IP = dyn_cast<InstrProfInstBase>(&Instr))
+        InstrProfInsts.push_back(IP);
```
Why do we need to hold the counter increments in intermediate vector now?

xur-llvm replied:
hmm. This is a good question. I have this  since the first implementation. We need this at that time as the CFG changed in the middle of the iterator.  Let me try to see if we still need it.

xur-llvm replied:
> hmm. This is a good question. I have this since the first implementation. We need this at that time as the CFG changed in the middle of the iterator. Let me try to see if we still need it.

We need to use a buffer to save the increment. I tried: Removing the buffer breaks the iterator, might due to the control flow introduced.

xur-llvm replied:
Ack. Will add a comment.

WenleiHe commented:
Does this also work with CS-IRPGO?

xur-llvm commented:
> Does this also work with CS-IRPGO?

Yes. This works with CS-IRPGO. There is a test for this.

xur-llvm committed:
Integrate David and Wenlei review commments.

xur-llvm commented:
Upload the patch that integrated comments from David and Wenlei.

xur-llvm commented on llvm/test/Transforms/PGOProfile/instrprof_simple_sampling.ll:
```diff
@@ -0,0 +1,39 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s --passes=instrprof --sampled-instrumentation --sampled-instr-burst-duration=1 --sampled-instr-period=1009 -S | FileCheck %s
```
Ack. Will add one.

xur-llvm committed:
[PGO] sampled instrumention

Add some comments and a test per Wenlei's suggestion.

xur-llvm commented:
Update the patch based on Wenlei's suggestion.

WenleiHe reviewed (approved):
lgtm, thanks.

xur-llvm committed:
Merge branch 'main' into main

ormris commented:
I'm seeing some buildbot failures after this commit: https://lab.llvm.org/buildbot/#/builders/66/builds/1956

@xur-llvm Could you take a look?

dyung commented:
Another buildbot showing a failure here: https://lab.llvm.org/buildbot/#/builders/174/builds/2103

xur-llvm commented:
Looking at the failures now.

-Rong

On Mon, Jul 22, 2024 at 10:54 AM dyung ***@***.***> wrote:

> Another buildbot showing a failure here:
> https://lab.llvm.org/buildbot/#/builders/174/builds/2103
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/69535#issuecomment-2243505449>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOI42XQGU4WVLIDG4B4QCETZNVBNHAVCNFSM6AAAAAA6GII2ZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENBTGUYDKNBUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ormris commented:
If some investigation is required, we would appreciate a revert.

xur-llvm commented:
I think this was due to compiler-rt/include/profile/InstrProfData.inc  not synced to llvm/include/llvm/ProfileData/InstrProfData.inc.

xur-llvm commented:
Create a PR to fix the issue: https://github.com/llvm/llvm-project/pull/99930
